R/r-session.R

```diff
@@ -0,0 +1,306 @@
+                     stdout = options$stdout, stderr = options$stderr,
+                     connections = list(private$pipe[[2]]))
+  )
+  private$started_at <- Sys.time()
+  private$state <- "starting"
+
+  ## Make child report back when ready
+  private$report_back(200, "ready to go")
+
+  invisible(self)
+}
+
+rs_run <- function(self, private, func, args, timeout) {
+  self$call(func, args)
+  self$wait_for_call(timeout)
+  self$get_result()
+}
+
+rs_call <- function(self, private, func, args) {
+  private$update_state()
+  if (private$state != "idle") stop("R session busy")
+
+  ## Save the function in a file
+  private$options$func <- func
+  private$options$args <- args
+  private$options$func_file <- save_function_to_temp(private$options)
+  private$options$result_file <- tempfile()
+  private$options$tmp_files <-
+    c(private$options$tmp_files, private$options$func_file,
+      private$options$result_file)
+
+  ## Run an expr that loads it, in the child process, with error handlers
+  expr <- make_vanilla_script_expr(private$options$func_file,
+                                   private$options$result_file,
+                                   private$options$error)
+  cmd <- paste0(deparse(expr), "\n")
+
+  ## Write this to stdin
+  private$write_for_sure(cmd)
+
+  ## Report back when done
+  report_str <- paste0("DONE", basename(private$options$result_file))
+  private$report_back(200, report_str)
+
+  private$state <- "busy"
+}
+
+rs_wait_for_call <- function(self, private, timeout) {
+  if (private$state %in% c("finished", "ready", "idle")) return()
+
+  pr <- poll(list(private$pipe[[1]]), timeout)[[1]]
+  if (pr == "ready") {
+    if (private$state == "starting") {
+      private$state <- "idle"
+    } else {
+      private$state <- "ready"
+    }
+    invisible(conn_read_lines(private$pipe[[1]], 1))
+  } else {
+    invisible()
+  }
+}
+
+rs_get_result <- function(self, private) {
+  if (private$state != "ready") private$update_state()
+
+  get_my_result <- function() {
+    ## This is artificial...
+    out <- list(
+      status = 0,
+      stdout = "",
+      stderr = "",
+      timeout = FALSE
+    )
+    res <- get_result(out, private$options)
+    private$state <- "idle"
+    unlink(private$options$tmp_files, recursive = TRUE)
+    private$options$tmp_files <- NULL
+    res
+  }
+
+  switch(
+    private$state,
+    "finished" = stop("R session already finished"),
+    "idle" = stop("R session is idle"),
+    "busy" = stop("R session still busy"),
+    "starting" = stop("R session still starting"),
+    "ready" = get_my_result()
+  )
+}
+
+rs_get_running_time <- function(self, private) {
+  now <- Sys.time()
+  c(total = now - private$started_at,
+    current = now - private$fun_started_at)
+}
+
+rs_get_state <- function(self, private) {
+  private$update_state()
+  private$state
+}
+
+rs_finish <- function(self, private, grace) {
+  close(self$get_input_connection())
+  self$poll_io(grace)
```

Comment:
Is 200 (ms?) enough time to ensure the everything closes gracefully?

Reply:
200ms, yes. Should be, unless the R session has `.Last` which takes longer to run. For this odd case, the grace period is user configurable.